docs/specifications/tests/Basic-TP/basic02.md

```diff
@@ -1,73 +1,220 @@
+B02_NS_NO_RESPONSE   | WARNING  | ns            | Name server "{ns}" does not respond to an SOA query.
+B02_UNEXPECTED_RCODE | ERROR    | ns, rcode     | Name server "{ns}" responds with an unexpected RCODE name ("{rcode}") on an SOA query.
+
+The value in the Level column is the default severity level of the message. The
+severity level can be changed in the [Zonemaster-Engine profile]. Also see the
+[Severity Level Definitions] document.
+
+The argument names in the Arguments column lists the arguments used in the
+message. The argument names are defined in the [argument list].
+
+
+## Test procedure
+
+In this section and unless otherwise specified below, the terms "[DNS Query]"
+follow the specification for DNS queries as specified in
+[DNS Query and Response Defaults]. The handling of the DNS responses on the DNS
+queries follow, unless otherwise specified below, what is specified for
+[DNS Response] in the same specification.
+
+
+1. Create a [DNS Query] with query type SOA and query name *Child Zone*
+   ("SOA Query").
+
+2. Create the following empty sets:
+   1. Name server name and IP ("Auth Response on SOA Query").
+   2. Name server name and IP ("Broken NS").
+   3. Name server name and IP ("NS not auth").
+   4. Name server name ("NS Cannot Resolve Into IP").
+   5. Name server name and IP ("No Response From NS").
+   6. Name server name and IP plus [RCODE Name] ("Unexpected RCODE").
+   7. Name server name and IP ("Delegation NS").
+
+3. Populate the set *Delegation NS* with name and IP address for the name
+   servers of the delegation of *Child Zone*.
+   1. If the test is an undelegated test, then:
+      1. Use *Undelegated NS*, *Undelegated Glue IP* and
+      *Undelegated Non-Glue IP*.
+      2. If any [out-of-bailiwick] name server name in the set has no IP address
+         then do a recursive lookup for address records (both IPv4 and IPv6) for
+         that name and add resolved addresses, if any, to the set.
+   2. Else, do:
+      1. Retrieve the NS records for *Child Zone* using [Method 2] and the IP
+         addresses ([glue records][glue record]) for any [in-bailiwick] name
+         servers using [Method 4].
+      2. Retrieve the IP addresses for any [out-of-bailiwick] name servers
+         using recursive lookup for address records (both IPv4 and IPv6) for
+         that name and add resolved addresses, if any, to the set.
+
+4. If the *Delegation NS* set is empty, then do:
+   1. Output *[B02_NO_DELEGATION]* with *Child Zone* name.
+   2. Exit these test procedures.
+
+5. Else, for each name server name in the *Delegation NS* set do:
+   1. If the name server name has no IP address then do:
+      1. Add the name server name to the *NS Cannot Resolve Into IP*.
+      2. Go to next name server name.
+   2. If the name server name has multiple IP addresses, then repeat the loop
+      and the following steps for each IP address, else go to next name server
+      name in the loop.
```

Comment:
Did you consider using a set of "Name server name and array of IPs" for the "Delegation NS" set? I think the procedure could be clearer with two loops instead of one.

Reply:
I kept the name, but I clarified and created a loop over the IP addresses as you suggested. Since we often have both IPv4 and IPv6 addresses it is actually a common case with more than one address.